contentcuration/contentcuration/utils/nodes.py

```diff
@@ -275,3 +286,189 @@ def generate_diff(updated_id, original_id):
+                WHERE contentnode_id IN ( SELECT id FROM contentcuration_contentnode WHERE ... )
+            ) subquery
+            ;
+
+        :return: An integer representing the resource size
+        """
+        sizes = self.queryset.values("checksum").distinct().aggregate(resource_size=Sum("file_size"))
+        return sizes['resource_size']
+
+    def modified_since(self, compare_datetime):
+        """
+        Determines if resources have been modified since ${compare_datetime}
+
+        SQL:
+            SELECT BOOL_OR(modified > ${compare_datetime})
+            FROM (
+                SELECT
+                    modified_at
+                FROM contentcuration_file
+                WHERE contentnode_id IN ( SELECT id FROM contentcuration_contentnode WHERE ... )
+            ) subquery
+            ;
+
+        :
+        :param compare_datetime: The datetime with which to compare.
+        :return: A boolean indicating whether or not resources have been modified since the datetime
+        """
+        compare_datetime = compare_datetime.isoformat() if isinstance(compare_datetime, datetime) else compare_datetime
+        result = self.queryset.aggregate(
+            modified_since=BoolOr(CombinedExpression(F('modified'), '>', Value(compare_datetime)))
+        )
+        return result['modified_since']
+
+
+STALE_MAX_CALCULATION_SIZE = 5000
+
+
+def calculate_resource_size(node, force=False):
+    """
+    Function that calculates the total file size of all files of the specified node and it's
+    descendants, if they're marked complete
+
+    :param node: The ContentNode for which to calculate resource size.
+    :param force: A boolean to force calculation if node is too big and would otherwise do so async
+    :return: A tuple of (size, stale)
+    :rtype: (int, bool)
+    """
+    cache = ResourceSizeCache(node)
+    db = ResourceSizeHelper(node)
+
+    size = None if force else cache.get_size()
+    modified = None if force else cache.get_modified()
+
+    # since we added file.modified as nullable, if the result is None/Null, then we know that it
+    # hasn't been modified since our last cached value, so we only need to check is False
+    if size is not None and modified is not None and db.modified_since(modified) is False:
+        # use cache if not modified since cache modified timestamp
+        return size, False
+
+    # if the node is too big to calculate its size right away, we return "stale"
+    if not force and node.get_descendant_count() > STALE_MAX_CALCULATION_SIZE:
+        return size, True
+
+    start = time.time()
+
+    # do recalculation, marking modified time before starting
+    now = timezone.now()
+    size = db.get_size()
+    cache.set_size(size)
+    cache.set_modified(now)
+    elapsed = time.time() - start
+
+    # log how long calculation took
+    msg = "Resource size recalculation for {} took {}s".format(node.pk, elapsed)
+    if elapsed > 5:
+        logging.warning(msg)
```

Comment:
Could do something similar to what we do in the sync endpoint here: https://github.com/learningequality/studio/blob/hotfixes/contentcuration/contentcuration/viewsets/sync/endpoint.py#L177 so that we get this logged in Sentry, so we're more likely to see it and take notice!

Reply:
Oh perfect, I'll update this